test/com/google/enterprise/gsafeed/GsafeedHelperTest.java

```diff
@@ -0,0 +1,270 @@
+        + "    <record url='http://www.example.com/hello01' "
+        + "      mimetype='text/plain'"
+        + "      last-modified='Tue, 6 Nov 2007 12:45:26 GMT'>"
+        + "      <content> &j; </content>"                            // <----
+        + "    </record>"
+        + "  </group>"
+        + "</gsafeed>";
+
+    externalEntityFeed =
+        "<?xml version='1.0' encoding='utf-8'?>"
+        + "<!DOCTYPE gsafeed PUBLIC '-//Google//DTD GSA Feeds//EN' '' [ "
+        + "<!ENTITY xxe SYSTEM '" + privateFile.toURI() + "'>"       // <----
+        + "]>"
+        + "<gsafeed>"
+        + "  <header>"
+        + "    <datasource>sample</datasource>"
+        + "    <feedtype>full</feedtype>"
+        + "  </header>"
+        + "  <group>"
+        + "    <record url='http://www.example.com/hello01' "
+        + "      mimetype='text/plain'"
+        + "      last-modified='Tue, 6 Nov 2007 12:45:26 GMT'>"
+        + "      <content> &xxe; </content>"                          // <----
+        + "    </record>"
+        + "  </group>"
+        + "</gsafeed>";
+
+    externalParameterEntityFeed =
+        "<?xml version='1.0' encoding='utf-8'?>"
+        + "<!DOCTYPE gsafeed PUBLIC '-//Google//DTD GSA Feeds//EN' '' [ "
+        + "<!ENTITY % dtd SYSTEM "
+        + "'" + evilDtd.toURI() + "'>"                               // <----
+        + "%dtd;"
+        + "%all;"
+        + "]>"
+        + "<gsafeed>"
+        + "  <header>"
+        + "    <datasource>sample</datasource>"
+        + "    <feedtype>full</feedtype>"
+        + "  </header>"
+        + "  <group>"
+        + "    <record url='http://www.example.com/hello01'"
+        + "      mimetype='text/plain'"
+        + "      last-modified='Tue, 6 Nov 2007 12:45:26 GMT'>"
+        + "      <content> &fileContents; </content>"                 // <----
+        + "    </record>"
+        + "  </group>"
+        + "</gsafeed>";
+  }
+
+  private static void writeFile(File file, String content) throws IOException {
+    FileWriter writer = new FileWriter(file);
+    try {
+      writer.write(content);
+    } finally {
+      writer.close();
+    }
+  }
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testWithDtd() throws Exception {
+    Gsafeed feed = GsafeedHelper.unmarshalWithDtd(
+        new ByteArrayInputStream(validFeed.getBytes("UTF-8")));
+    assertEquals("sample", feed.getHeader().getDatasource());
+  }
+
+  @Test
+  public void testWithoutDtd() throws Exception {
+    Gsafeed feed = GsafeedHelper.unmarshalWithoutDtd(
+        new ByteArrayInputStream(validFeed.getBytes("UTF-8")));
+    assertEquals("sample", feed.getHeader().getDatasource());
+  }
+
+  @Test
+  public void testWithDtdInvalidDoc() throws Exception {
+    thrown.expect(SAXParseException.class);
+    thrown.expectMessage(
+        "Element type \"invalid-element\" must be declared.");
+    Gsafeed feed = GsafeedHelper.unmarshalWithDtd(
```

Comment:
I think we usually leave off the local variable in this case, since the initialization never happens.